File: Problems/No.160_Intersection_of_Two_Linked_Lists/intersection_of_two_linked_lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        
        
        # count the length of linked list
        def CountNodes(node):
            count = 1
            while (node.next):
                count += 1
                node = node.next
            return count
        
        # parameters
        A_num, B_num = CountNodes(headA), CountNodes(headB)
        a, b = headA, headB
        
        # preprocess for matching the list length
        for i in range(A_num-B_num):
            a = a.next
        for i in range(B_num-A_num):
            b = b.next
        
        # start checking
        while (a != b):
            a = a.next
            b = b.next
        
        return a

File: Problems/No.160_Intersection_of_Two_Linked_Lists/intersection_of_two_linked_lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    
    // count the number of nodes
    int CountNodes(ListNode *node) {
        
        int count = 1;
        while (node->next) {
            
            count++;
            node = node->next;
            
        }
        
        return count;
        
    }
    
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
     
        // parameters
        int A_num = CountNodes(headA);
        int B_num = CountNodes(headB);
        ListNode *a = headA, *b = headB;
        
        // preprocess to match the list length
        for (int i=0; i<A_num-B_num; i++) a = a->next;
        for (int i=0; i<B_num-A_num; i++) b = b->next;
        
        // start checking
        while (a != b) {
            
            a = a->next;
            b = b->next;
            
        }
        
        return a;
        
    }
};
